test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc

```diff
@@ -0,0 +1,185 @@
+  bool health_checks_;
+  std::string inner_socket_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ProxyProtocolIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// Test sending proxy protocol v1
+TEST_P(ProxyProtocolIntegrationTest, TestV1ProxyProtocol) {
+  setup(envoy::config::core::v3::ProxyProtocolConfig::V1, false,
+        "envoy.transport_sockets.raw_buffer");
+  initialize();
+
+  auto listener_port = lookupPort("listener_0");
+  auto tcp_client = makeTcpConnection(listener_port);
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection_));
+
+  std::string observed_data;
+  ASSERT_TRUE(tcp_client->write("data"));
+  if (GetParam() == Network::Address::IpVersion::v4) {
+    ASSERT_TRUE(fake_upstream_connection_->waitForData(48, &observed_data));
+    EXPECT_THAT(observed_data, testing::StartsWith("PROXY TCP4 127.0.0.1 127.0.0.1 "));
+  } else if (GetParam() == Network::Address::IpVersion::v6) {
+    ASSERT_TRUE(fake_upstream_connection_->waitForData(36, &observed_data));
+    EXPECT_THAT(observed_data, testing::StartsWith("PROXY TCP6 ::1 ::1 "));
+  }
+  EXPECT_THAT(observed_data, testing::EndsWith(absl::StrCat(" ", listener_port, "\r\ndata")));
+
+  auto previous_data = observed_data;
+  observed_data.clear();
+  ASSERT_TRUE(tcp_client->write(" more data"));
+  ASSERT_TRUE(fake_upstream_connection_->waitForData(previous_data.length() + 10, &observed_data));
+  EXPECT_EQ(previous_data + " more data", observed_data);
+
+  tcp_client->close();
+  ASSERT_TRUE(fake_upstream_connection_->waitForDisconnect());
+}
+
+// Test header is sent unencrypted using a TLS inner socket
+TEST_P(ProxyProtocolIntegrationTest, TestTLSSocket) {
+  setup(envoy::config::core::v3::ProxyProtocolConfig::V1, false, "envoy.transport_sockets.tls");
+  initialize();
+
+  auto listener_port = lookupPort("listener_0");
+  auto tcp_client = makeTcpConnection(listener_port);
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection_));
+
+  ASSERT_TRUE(tcp_client->write("data"));
+  if (GetParam() == Network::Address::IpVersion::v4) {
+    ASSERT_TRUE(fake_upstream_connection_->waitForData(
+        fake_upstream_connection_->waitForInexactMatch("PROXY TCP4 127.0.0.1 127.0.0.1 ")));
+  } else if (GetParam() == Network::Address::IpVersion::v6) {
+    ASSERT_TRUE(fake_upstream_connection_->waitForData(
+        fake_upstream_connection_->waitForInexactMatch("PROXY TCP6 ::1 ::1 ")));
+  }
+
+  tcp_client->close();
+  ASSERT_TRUE(fake_upstream_connection_->waitForDisconnect());
+}
+
+// Test sending proxy protocol health check
+TEST_P(ProxyProtocolIntegrationTest, TestProxyProtocolHealthCheck) {
+  setup(envoy::config::core::v3::ProxyProtocolConfig::V1, true,
+        "envoy.transport_sockets.raw_buffer");
+  initialize();
+
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection_));
```

Comment:
Would be nice to verify the actual header from the health check, but I was just getting timeouts in `waitForData`, and it showed no data. Not sure if there's a better way to test this?

Reply:
again maybe -l trace or cerr logging?   It may be easier to snag values from other health checking tests in tcp_proxy_integration_test.cc